src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Author: Tej Sukhatme
+ */
+
+#include <shogun/features/DotFeatures.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/labels/RegressionLabels.h>
+#include <shogun/lib/observers/ObservedValueTemplated.h>
+#include <shogun/machine/GLM.h>
+#include <shogun/machine/LinearMachine.h>
+#include <shogun/mathematics/NormalDistribution.h>
+#include <shogun/mathematics/RandomNamespace.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/optimization/ConstLearningRate.h>
+#include <shogun/optimization/ElasticNetPenalty.h>
+#include <shogun/optimization/GradientDescendUpdater.h>
+#include <shogun/optimization/SGDMinimizer.h>
+
+#include <cmath>
+
+using namespace shogun;
+
+GLM::GLM()
+{
+	SG_ADD(
+	    &distribution, "distribution_type",
+	    "variable to store name of distribution type",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_eta, "eta",
+	    "threshold parameter that linearizes the exp() function above eta",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_lambda, "lambda", "regularization parameter of penalty term",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_alpha, "alpha",
+	    "weighting between L1 penalty and L2 penalty term of the loss function",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_tolerance, "tolerance", "convergence threshold or stopping criteria",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_learning_rate, "learning_rate", "learning rate for gradient descent",
+	    ParameterProperties::HYPER);
+
+	m_gradient_updater = std::make_shared<GradientDescendUpdater>();
+	// m_learning_rate = std::make_shared<ConstLearningRate>();
+	m_penalty = std::make_shared<ElasticNetPenalty>();
+	m_cost_function = std::make_shared<GLMCostFunction>();
+}
+
+GLM::GLM(
+    GLM_DISTRIBUTION distr, float64_t alpha, float64_t lambda,
+    float64_t learning_rate, int32_t max_iterations, float64_t tolerance,
+    float64_t eta)
+    : GLM()
+{
+	distribution = distr;
+	m_alpha = alpha;
+	m_lambda = lambda;
+	m_learning_rate = learning_rate;
+	m_max_iterations = max_iterations;
+	m_tolerance = tolerance;
+	m_eta = eta;
+
+	// m_learning_rate->set_const_learning_rate(learning_rate);
+
+	m_penalty->set_l1_ratio(m_alpha);
+}
+
+std::shared_ptr<RegressionLabels>
+GLM::apply_regression(std::shared_ptr<Features> data)
+{
+	LinearMachine::set_features(std::static_pointer_cast<DotFeatures>(data));
+
+	if (!LinearMachine::features)
+		return std::make_shared<RegressionLabels>(SGVector<float64_t>());
```

Comment:
these two lines can just be `require(features)`, I think that is how other algorithms do this

Reply:
I did it this way as it is done this way here: 
https://github.com/shogun-toolbox/shogun/blob/ec557d8604b7bf5c2f4ce411aaa2dc0970839d8d/src/shogun/machine/LinearMachine.cpp#L78-L79
should I just use `require` instead?

Reply:
imo an error explaining the issue is better than an empty result

Reply:
Right, I'll do that.